File: projecteuler/euler009.rb
MAX = 3001
squares = Array.new(MAX)
i = 1
while i < MAX do
  squares[i] = i * i
  i += 1
end

gets.to_i.times do
  n = gets.to_i
  max_product = -1

  1.upto(n / 3 + 1) do |a|
    next if 2 * n - 2 * a == 0
    b = (squares[n] - 2 * n * a) / (2 * n - 2 * a)
    c = n - a - b
    if c > 0 && c > b && c > a && b > a && squares[a] + squares[b] == squares[n - a - b]
      max_product = [a * b * c, max_product].max
    end
  end

  puts max_product
end

File: projecteuler/Euler009.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {
  public static final int MAX = 3001;

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int i, n, product, maxProduct, a, b, c;
    int[] squares = new int[MAX];
    
    for(i = 1; i < MAX; i++) {
      squares[i] = i * i;
    }

    int t = scanner.nextInt();
    for (int k = 0; k < t; k++) {
      maxProduct = -1;
      n = scanner.nextInt();
      for (a = 1; a < n / 3 + 1; a++) {
        if (2 * n - 2 * a == 0) {
          continue;
        }
        b = (squares[n] - 2 * n * a) / (2 * n - 2 * a);
        c = n - a - b;
        if (c > 0 && c > b && c > a && b > a && squares[a] + squares[b] == squares[c]) {
          product = a * b * c;
          if (product > maxProduct) {
            maxProduct = product;
          }
        }
      }
      System.out.println(maxProduct);
    }
  }
}
